dev/archery/archery/benchmark/google.py

```diff
@@ -0,0 +1,193 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from itertools import filterfalse, groupby, tee
+import json
+import subprocess
+
+from .core import Benchmark
+from ..utils.command import Command
+from ..utils.git import git
+
+
+def partition(pred, iterable):
+    # adapted from python's examples
+    t1, t2 = tee(iterable)
+    return list(filter(pred, t1)), list(filterfalse(pred, t2))
+
+
+class GoogleBenchmarkCommand(Command):
+    def __init__(self, benchmark_bin, benchmark_filter=None):
+        self.bin = benchmark_bin
+        self.benchmark_filter = benchmark_filter
+
+    def list_benchmarks(self):
+        argv = ["--benchmark_list_tests"]
+        if self.benchmark_filter:
+            argv.append(f"--benchmark_filter={self.benchmark_filter}")
+        result = self.run(*argv, stdout=subprocess.PIPE,
+                          stderr=subprocess.PIPE)
+        return [b for b in result.stdout]
+
+    def results(self):
+        argv = ["--benchmark_format=json", "--benchmark_repetitions=20"]
+
+        if self.benchmark_filter:
+            argv.append(f"--benchmark_filter={self.benchmark_filter}")
+
```

Comment:
It'd be useful to provide some progress output as each test is run so users know nothing is hung.

Maybe benchmarks could be run one at a time with messages naming each?

Reply:
Feel free to commit, but it would requires some more thinking:

1. Rework how to capture results from google benchmark (right now from stdout). We can use `--benchmark_output`, then we'll get "progress" in stdout.

2. `archery` stdout is now clobbered with this result, so either we redirect the previous point output into stderr, or into the logger.

I'm not very satisfied with either answer. Note that in all cases, you can get some feedback with `--debug`.

Reply:
One way to do it would be:

```suggestion
    @property
    def suite_name(self):
        return os.path.splitext(os.path.basename(self.bin))[0]

    def results(self):
        argv = ["--benchmark_format=json", "--benchmark_repetitions=20"]

        results = { "benchmarks": [] }
        for name in self.list_benchmarks():
            print(f"running {self.suite_name}.{name}")
            result = json.loads(self.run(*argv, f"--benchmark_filter={name}",
                                         stdout=subprocess.PIPE,
                                         stderr=subprocess.PIPE).stdout)
            results["context"] = result["context"]
            results["benchmarks"] += result["benchmarks"]

        return results
```

re stdout clobbering: the output already seems clobbered by things like 'ninja: no work to do.`

Maybe it would be better to provide the option to specify filenames for comparison (and/or benchmark) output json, rather than rely on stdio?